File: wbdv-ngo-sp21-server-java/src/main/java/com/example/wbdvngosp21serverjava/services/WidgetService.java
package com.example.wbdvngosp21serverjava.services;

import org.springframework.stereotype.Service;
import com.example.wbdvngosp21serverjava.models.Widget;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class WidgetService {
  private List<Widget> widgets;
  private Integer currentIndex;

  public WidgetService() {
    widgets = new ArrayList<Widget>();
    currentIndex = 0;
  }

  public Widget createWidget(String tid, Widget widget) {
    widget.setTopicId(tid);
    widget.setId(currentIndex);
    currentIndex++;

    widgets.add(widget);
    return widget;
  }

  public List<Widget> findWidgetsForTopic(String tid) {
    return widgets.stream().filter(
            widget -> widget.getTopicId().equals(tid)).collect(Collectors.toList());
  }

  public int updateWidget(String wid, Widget widget) {
    Integer id = Integer.parseInt(wid);

    for (int i = 0; i < widgets.size(); i++) {
      if (widgets.get(i).getId().equals(id)) {
        widgets.set(i, widget);
        return 1;
      }
    }
    return 0;
  }

  public int deleteWidget(String wid) {
    Integer id = Integer.parseInt(wid);

    for (int i = 0; i < widgets.size(); i++) {
      if (widgets.get(i).getId().equals(id)) {
        widgets.remove(i);
        return 1;
      }
    }
    return 0;
  }

  public List<Widget> findAllWidgets() {
    return widgets;
  }

  public Widget findWidgetById(String wid) {
    Integer id = Integer.parseInt(wid);

    for (int i = 0; i < widgets.size(); i++) {
      if (widgets.get(i).getId().equals(id)) {
        return widgets.get(i);
      }
    }
    return null;
  }

}

File: wbdv-ngo-sp21-client-react/src/services/widget-service.js
const API_URL = "http://localhost:8080/api"

export const createWidget = (tid, widget) => {
    return fetch(`${API_URL}/topics/${tid}/widgets`, {
        type: 'POST',
        body: JSON.stringify(widget),
        headers: {
            'content-type': 'application/json'
        }}).then((response) => response.json());
}

export const findWidgetById = wid => {
    return fetch(`${API_URL}/widgets/${wid}`)
        .then(response => response.json());
}

export const findWidgetsForTopic = tid => {
    return fetch(`${API_URL}/topics/${tid}/widgets`)
        .then(response => response.json());
}

export const findAllWidgets = () => {
    return fetch(`${API_URL}/widgets`)
        .then(response => response.json());
}

export const updateWidget = (wid, widget) => {
    return fetch(`${API_URL}/widgets/${wid}`, {
        type: 'PUT',
        body: JSON.stringify(widget),
        headers: {
            'content-type': 'application/json'
        }
    }).then(response => response.json());
}

export const deleteWidget = wid => {
    return fetch(`${API_URL}/widgets/${wid}`, {
        type: 'DELETE'
    }).then(response => response.json());
}

export default {
    createWidget,
    findWidgetById,
    findWidgetsForTopic,
    findAllWidgets,
    updateWidget,
    deleteWidget
}